README.md
# ZfrOAuth2Server

[![Build Status](https://travis-ci.org/zf-fr/zfr-oauth2-server.png)](https://travis-ci.org/zf-fr/zfr-oauth2-server)
[![Latest Stable Version](https://poser.pugx.org/zfr/zfr-oauth2-server/v/stable.png)](https://packagist.org/packages/zfr/zfr-oauth2-server)
[![Coverage Status](https://coveralls.io/repos/github/zf-fr/zfr-oauth2-server/badge.svg?branch=master)](https://coveralls.io/github/zf-fr/zfr-oauth2-server?branch=master)
[![Total Downloads](https://poser.pugx.org/zfr/zfr-oauth2-server/downloads.png)](https://packagist.org/packages/zfr/zfr-oauth2-server)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/prolic/zfr-oauth2-server)

ZfrOAuth2Server is a PHP library that implements the OAuth 2 specification. It's main goal is to be a clean, PHP 7.0+
library that aims to be used with any persistence layer of choice. It is compatible with
[PSR-7](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md) request and responses which makes
it possible to use with any framework compatible with PSR-7.

Currently, ZfrOAuth2Server does not implement the whole specification (implicit grant is missing), so you are encouraged to have a look at the doc if ZfrOAuth2Server can be used in your application. 

However, it implements the additional [token revocation](https://tools.ietf.org/html/rfc7009) specification.

Here are other OAuth2 library you can use:

- [OAuth2 Server from PHP-League](https://github.com/php-loep/oauth2-server)
- [OAuth2 Server from Brent Shaffer](https://github.com/bshaffer/oauth2-server-php)

## Requirements

- PHP 7.2 or higher

## To-do

- Write documentation
- Security audit
- Review of the whole spec
- Testing the authorization server more extensively
- Add implicit grant

## Versioning note

Please note that until we reach 1.0, we **WILL NOT** follow semantic version. This means that BC can occur between
0.1.x and 0.2.x releases.

The current pre release of a completely rewritten version, is it not copatible with the previous implementation - which is considered EOL - see the [legacy-0.7](https://github.com/zf-fr/zfr-oauth2-server/tree/legacy-0.7) branch. 

See the [CHANGELOG](CHANGELOG.md)

## Installation

use Composer to install:

```sh
php composer.phar require zfr/zfr-oauth2-server:^0.9-beta
```

## Support

- File issues at [https://github.com/zf-fr/zfr-oauth2-server/issues](https://github.com/zf-fr/zfr-oauth2-server/issues).
- Say hello in our [gitter](https://gitter.im/prolic/zfr-oauth2-server) chat.

## Configuration

Several Apache modules will strip HTTP authorization headers such as `Authorization` to try to enhance security by preventing scripts from seeing sensitive information unless the developer explicitly enables this.

Many of these modules will allow such headers if you simply add the following line to .htaccess (or the vhost directory directive).

```
CGIPassAuth on
```
since: [Apache 2.4.13](https://httpd.apache.org/docs/trunk/mod/core.html#cgipassauth)


## Documentation

ZfrOAuth2Server is based on the [RFC 6749](http://tools.ietf.org/html/rfc6749) documentation.

### Why use OAuth2?

OAuth2 is an authentication/authorization system that allows that can be used to:

* Implement a stateless authentication mechanism (useful for API)
* Allow third-party to connect to your application securely
* Securing your application through the use of scopes

OAuth2 is a dense, extensible specification that can be used for a wide number of use-cases. As of today,
ZfrOAuth2Server implements three of the four official grants: AuthorizationGrant, ClientCredentialsGrant, PasswordGrant. Additionally a RefreshTokenGrant is provided to obtain new access tokens. ImplicitGrant and JWTTokens are forthcoming (help wanted).

### How OAuth2 works?

This documentation does not aim to explain in details how OAuth2 work. Here is [a nice resource](http://aaronparecki.com/articles/2012/07/29/1/oauth2-simplified) you can read. However, here is the basic idea of how OAuth2 works:

1. A resource owner (your JavaScript API, your mobile application...) asks for a so-called "access token" to an
 authorization server. There are several strategies that depends on the use-case. Those strategies are called
 "grants". For instance, the "password grant" assumes that the resource owner sends its username/password. In all
 cases, your authorization server responds with an access token (and an optional refresh token).
2. The client sends this access token to each request that is made to your API. It is used by a "resource server"
to map this access token to a user in your system.

Choosing the grant type depends on your application. Here are a few hints about which one to choose:

* If you are the only consumer of your API (for instance, your JavaScript application make calls to your API), you
should use the "password grant". Because you trust your application, it is not a problem to send username/password.
* If you want a third-party code to connect to your API, and that you are sure that this third-party can keep secrets
(this means the client is not a JavaScript API, or a mobile application): you can use the client credentials grant.
* If you want third-party code to connect to your API, and that those third-party applications cannot keep secret
(think about an unofficial Twitter client that connect to your Twitter account, for instance), you should use the
authorization grant.

### Using the authorization server

The authorization server goal is to accept a request, and generate token. An authorization server can deny a
request (for instance, if parameters are missing, or if username/password are incorrect).

To use an authorization server, you must first decide which grant you want to support. Some applications should
only support one type of grant, others may support all of the available grant. This is completely up to you, and
you should have a solid understanding of all those grants first. For instance, here is how you would create an
authorization server that support the authorization only:

```php
$authTokenService    = new TokenService($objectManager, $authTokenRepository, $scopeRepository);
$accessTokenService  = new TokenService($objectManager, $accessTokenRepository, $scopeRepository);
$refreshTokenService = new TokenService($objectManager, $refreshTokenRepository, $scopeRepository);

$authorizationGrant  = new AuthorizationGrant($authTokenService, $accessTokenService, $refreshTokenService);
$authorizationServer = new AuthorizationServer([$authorizationGrant]);

// Response contains the various parameters you can return
$response = $authorizationServer->handleRequest($request);
```

The request must be a valid `Psr\Http\Message\ServerRequestInterface`, and the authorization server returns a `Psr\Http\Message\ResponseInterface` object
that is compliant with the OAuth2 specification.

#### Passing a user

Most of the time, you want to associate an access token to a user. This is the only way to map a token to a user
of your system. To do this, you can pass an optional second parameter to the `handleRequest`. This class must
implements the `ZfrOAuth2\Server\Model\TokenOwnerInterface` interface:

```php
$user = new User(); // must implement TokenOwnerInterface

// ...

$response = $authorizationServer->handleRequest($request, $user);
```

The AuthorizationServerMiddleware is able to do this for you and retrieve a user instance from a (configurable) request attribute. It is up to you to provide middleware which runs with a higher priority to add a TokenOwnerInterface instance to the request attribute. 

Example of such a implementation which uses LaminasAuthentication and a TemplateRenderer from Mezzio.

```
final class OAuth2AuthorizationFlow
{
    /**
     * @var AuthenticationService
     */
    private $authenticationService;

    /**
     * @var ClientService
     */
    private $clientService;

    /**
     * @var TemplateRendererInterface
     */
    private $template;

    public function __construct(
        AuthenticationService $authenticationService,
        ClientService $clientService,
        TemplateRendererInterface $template
    ) {
        $this->authenticationService = $authenticationService;
        $this->clientService         = $clientService;
        $this->template              = $template;
    }

    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $out = null)
    {
        if ($this->authenticationService->hasIdentity()) {
            $request = $request->withAttribute('owner', $this->authenticationService->getIdentity());
        }

        if ($request->getMethod() === 'POST') {
            $post     = $request->getParsedBody();
            $approved = filter_var($post['approved'], FILTER_VALIDATE_BOOLEAN);

            if ($approved) {
                return $out($request, $response);
            }
        }

        $data  = [];
        $query = $request->getUri()->getQuery();
        parse_str($query, $data['query']);

        $data['client'] = $this->clientService->getClient($data['query']['client_id']);

        return new HtmlResponse($this->template->render('app::oauth2/authorize-request', $data));
    }
}
```

#### Revoking a token 

ZfrOAuth2Server supports revoking access and refresh tokens using the [RFC 7009 specification](https://tools.ietf.org/html/rfc7009).
You can use the `handleRevocationRequest` method in the AuthorizationServer. You must pass the following two POST parameters:

* `token`: the token to remove (either access or refresh token)
* `token_hint_type`: must be either `access_token` or `refresh_token` to indicate the authorization server which token
type to revoke.

If you need to revoke a token that was issued for a non-public client (this means a client that has a secret key), then you
MUST authenticate the request using the client id and secret.

> If you try to revoke a token that does not exist, it will return 200 SUCCESS request, according to the spec. However,
if the token is valid, but cannot be deleted for any reason (database is down...), then it returns a 503 SERVICE UNAVAILABLE
error!

### Using the resource server

You can use the resource server to retrieve the access token (by automatically extracting the data from the HTTP
headers). You can also specify scope constraints when retrieving the token:

```php
$accessTokenService = new TokenService($objectManager, $accessTokenRepository, $scopeRepository);
$resourceServer     = new ResourceServer($accessTokenService);

if (!$token = $resourceServer->getAccessToken($request, ['write']) {
    // there is either no access token, or the access token is expired, or the access token does not have
    // the `write` scope
}
```

The ResourceServerMiddleware is able to do this for you, simply have it run before any other middleware.

Example mezzio expressive route configuration.

```
[
            'name'            => 'command::commerce::create-store',
            'path'            => '/commerce/create-store',
            'middleware'      => [
                ResourceServerMiddleware::class,
                MyActionMiddleware::class,
            ],
            'allowed_methods' => ['OPTIONS', 'POST'],
        ],
```         

### Persistence layer

As of version 0.8-beta1 ZfrOAuth2Server has been rewritten to be persistence layer agnostic. Meaning it can by used with any prefered persistence layer.

Currently these packages provide a persistence layer;

- [ZfrOAuth2ServerDoctrine](https://github.com/zf-fr/zfr-oauth2-server-doctrine) for Doctrine 2


CHANGELOG.md
# Changelog

## dev-master

## v0.8.0-beta3

* Implements [PSR-15](https://github.com/php-fig/fig-standards/tree/master/proposed/http-middleware)

## v0.8.0-beta2

* When an token can't be found the returned error response by the resource server middleware is now in a similar format to other errors. This might BC if your client depends on the error key in the message.
* Added an server option so the request attribute for tokens can be configured

## v0.8.0-beta1

BC! Pre release of a completely rewritten library. It focusses on core OAuth2 functionality and has been decoupled from persistence. If you still need the previous implementation - which is considered EOL - see the [legacy-0.7](https://github.com/zf-fr/zfr-oauth2-server/tree/legacy-0.7) branch

* PHP7+ only
* 100% test coverage
* Uses [Zend\Diactoros](https://github.com/zendframework/zend-diactoros) to generate  [PSR-7 (Http Message)](https://github.com/php-fig/http-message) implementation.
* Uses [PSR-11 (Container)](https://github.com/php-fig/container) for dependency injection containers.
* Eventing has been removed
* Persistence has been decoupled, see our doctrine integration [ZfrOAuth2ServerDoctrine](https://github.com/zf-fr/zfr-oauth2-server-doctrine)
* Provides 5 Services
	* ZfrOAuth2\Server\Service\AccessTokenService
	* ZfrOAuth2\Server\Service\AuthorizationCodeService
	* ZfrOAuth2\Server\Service\ClientService
	* ZfrOAuth2\Server\Service\RefreshTokenService
	* ZfrOAuth2\Server\Service\ScopeService
* Provides 4 PSR7 Middleware's which are really nice but optional
	* ZfrOAuth2\Server\AuthorizationServerMiddleware
	* ZfrOAuth2\Server\ResourceServerMiddleware
	* ZfrOAuth2\Server\RevocationRequestMiddleware
	* ZfrOAuth2\Server\TokenRequestMiddleware

## v0.7.1

* Now properly triggers an `EVENT_CODE_CREATED` event instead of `EVENT_CODE_FAILED` when response is between 200 and 399 (previously, 
as 302 Redirect used to trigger a failed event, although it created an authorization code).

## v0.7.0

* [BC] PHP minimum version has been bumped to 5.5. As a consequence, Zend\Crypt dependency has been removed as some of 
features are built-in into PHP 5.5.
  
* [BC] Instead of Zend\Http requests and responses, the module now uses PSR7 requests and responses, for increased 
compatibility. If you are using the ZF2 module, this should be completely transparent to you.
  
* [BC] Contrary to Zend\Http requests and responses, PSR7 are stateless. If you are using events to modify the response, 
you will need to use a different way.
  
In ZfrOAuth2Server 0.6:

```php
public function tokenCreated(TokenEvent $event)
{
    // We can log the access token
    $accessToken = $event->getAccessToken();
    // ...
  
    // Or we can alter the response body, if we need to
    $body                 = $event->getResponseBody();
    $body['custom_field'] = 'bar';
  
    // Update the body
    $event->setResponseBody($body);
}
```

In ZfrOAuth2Server 0.7+:

```php
public function tokenCreated(TokenEvent $event)
{
    // Get the response
    $response = $event->getResponse();
    
    // ...
  
    // Response is a PSR-7 compliant response, so you modify it
    $response = $response->withHeader(...);
  
    // Do not forget to set back the response, as PSR-7 are immutable
    $event->setResponse($response);
}
```

* Interfaces for ResourceServer and AuthorizationServer has been added, for easier testing.
  
## v0.6.0

* In previous versions, ZfrOAuth2 would trigger an "InvalidAccessTokenException" if you'd try to call the `getToken` 
when no token was specified in either Authorization header or query param. Now, ZfrOAuth2 will simply return null 
(because no token was explicitly set). However, this exception will be trigger IF an access token is indeed given, but 
does not exist in your database, is expired or does not match scopes.

## v0.5.0

* Support for token revocation by implementing [RFC7009 specification](https://tools.ietf.org/html/rfc7009)

## v0.4.0

* Allow multiple redirect URI for client (there is a minor table schema change, as a consequence)
* Fix a potential security issue by being more restrictive on the redirect URI when creating an authorization code. Now, 
if someone send a custom redirect_uri in the query params, the OAuth2 server will first check if the given redirect URI is 
in the list of the authorized redirect URIs by the client. If that's not the case, an InvalidRequest exception will be 
returned, and no authorization code will be generated.
  
## v0.3.0

* Add support for the ZF2 event manager. You can now attach listeners that are called whenever a new authorization code is 
created or failed, or when a new access token is created or failed.

## v0.2.0

* [BC] The `isRequestValid` from the ResourceServer is now gone in favour of a simpler approach: you just need to call 
the `getAccessToken` from the ResourceServer (with optional scopes), and null will be returned if the token is either expired, does 
not exist or does not match given scopes.

## v0.1.1

* Tokens do not contain \ and / characters anymore (as it can lead to problems when the token is passed as a query param).

## v0.1.0

* First release!


src/AuthorizationServer.php
<?php

declare(strict_types = 1);

/*
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * This software consists of voluntary contributions made by many individuals
 * and is licensed under the MIT license.
 */

namespace ZfrOAuth2\Server;

use Laminas\Diactoros\Response;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Throwable;
use ZfrOAuth2\Server\Exception\OAuth2Exception;
use ZfrOAuth2\Server\Grant\AuthorizationServerAwareInterface;
use ZfrOAuth2\Server\Grant\GrantInterface;
use ZfrOAuth2\Server\Model\Client;
use ZfrOAuth2\Server\Model\TokenOwnerInterface;
use ZfrOAuth2\Server\Service\AccessTokenService;
use ZfrOAuth2\Server\Service\ClientService;
use ZfrOAuth2\Server\Service\RefreshTokenService;

/**
 * The authorization server main role is to create access tokens or refresh tokens
 *
 * @author  Michaël Gallego <mic.gallego@gmail.com>
 * @licence MIT
 */
class AuthorizationServer implements AuthorizationServerInterface
{
    /**
     * @var ClientService
     */
    private $clientService;

    /**
     * A list of grant
     *
     * @var GrantInterface[]
     */
    private $grants = [];

    /**
     * A list of grant that can answer to an authorization request
     *
     * @var GrantInterface[]
     */
    private $responseTypes = [];

    /**
     * @var AccessTokenService
     */
    private $accessTokenService;

    /**
     * @var RefreshTokenService
     */
    private $refreshTokenService;

    /**
     * @param ClientService       $clientService
     * @param GrantInterface[]    $grants
     * @param AccessTokenService  $accessTokenService
     * @param RefreshTokenService $refreshTokenService
     */
    public function __construct(
        ClientService $clientService,
        array $grants,
        AccessTokenService $accessTokenService,
        RefreshTokenService $refreshTokenService
    ) {
        $this->clientService = $clientService;
        $this->accessTokenService = $accessTokenService;
        $this->refreshTokenService = $refreshTokenService;

        foreach ($grants as $grant) {
            if ($grant instanceof AuthorizationServerAwareInterface) {
                $grant->setAuthorizationServer($this);
            }

            $this->grants[$grant->getType()] = $grant;

            if ($responseType = $grant->getResponseType()) {
                $this->responseTypes[$responseType] = $grant;
            }
        }
    }

    /**
     * Check if the authorization server supports this grant
     */
    public function hasGrant(string $grantType): bool
    {
        return isset($this->grants[$grantType]);
    }

    /**
     * Get the grant by its name
     *
     * @throws OAuth2Exception (unsupported_grant_type) When grant type is not registered
     */
    public function getGrant(string $grantType): GrantInterface
    {
        if ($this->hasGrant($grantType)) {
            return $this->grants[$grantType];
        }

        // If we reach here... then no grant was found. Not good!
        throw OAuth2Exception::unsupportedGrantType(sprintf(
            'Grant type "%s" is not supported by this server',
            $grantType
        ));
    }

    /**
     * Check if the authorization server supports this response type
     */
    public function hasResponseType(string $responseType): bool
    {
        return isset($this->responseTypes[$responseType]);
    }

    /**
     * Get the response type by its name
     *
     * @throws OAuth2Exception (unsupported_grant_type) When response type is not registered
     */
    public function getResponseType(string $responseType): GrantInterface
    {
        if ($this->hasResponseType($responseType)) {
            return $this->responseTypes[$responseType];
        }

        // If we reach here... then no grant was found. Not good!
        throw OAuth2Exception::unsupportedResponseType(sprintf(
            'Response type "%s" is not supported by this server',
            $responseType
        ));
    }

    /**
     * @throws OAuth2Exception (invalid_request) If no "response_type" could be found in the GET parameters
     * @throws OAuth2Exception (invalid_clientt) If no client could be authenticated
     */
    public function handleAuthorizationRequest(
        ServerRequestInterface $request,
        TokenOwnerInterface $owner = null
    ): ResponseInterface {
        try {
            $queryParams = $request->getQueryParams();
            $responseType = $queryParams['response_type'] ?? null;

            if (null === $responseType) {
                throw OAuth2Exception::invalidRequest('No grant response type was found in the request');
            }

            $responseType = $this->getResponseType((string) $responseType);
            $client = $this->getClient($request, $responseType->allowPublicClients());

            if (null === $client) {
                throw OAuth2Exception::invalidClient('No client could be authenticated');
            }

            $response = $responseType->createAuthorizationResponse($request, $client, $owner);
        } catch (OAuth2Exception $exception) {
            $response = $this->createResponseFromOAuthException($exception);
        }

        return $response->withHeader('Content-Type', 'application/json');
    }

    /**
     * @throws OAuth2Exception (invalid_request) If no "grant_type" could be found in the POST parameters
     */
    public function handleTokenRequest(
        ServerRequestInterface $request,
        TokenOwnerInterface $owner = null
    ): ResponseInterface {
        $postParams = $request->getParsedBody();

        try {
            $grant = $postParams['grant_type'] ?? null;

            if (null === $grant) {
                throw OAuth2Exception::invalidRequest('No grant type was found in the request');
            }

            $grant = $this->getGrant((string) $grant);
            $client = $this->getClient($request, $grant->allowPublicClients());

            $response = $grant->createTokenResponse($request, $client, $owner);
        } catch (OAuth2Exception $exception) {
            $response = $this->createResponseFromOAuthException($exception);
        }

        // According to the spec, we must set those headers (http://tools.ietf.org/html/rfc6749#section-5.1)
        return $response->withHeader('Content-Type', 'application/json')
                        ->withHeader('Cache-Control', 'no-store')
                        ->withHeader('Pragma', 'no-cache');
    }

    /**
     * @throws OAuth2Exception (invalid_request) If no "token" is present
     * @throws OAuth2Exception (unsupported_token_type) If "token" is unsupported
     * @throws OAuth2Exception (invalid_client) If "token" was issued for another client and cannot be revoked
     */
    public function handleRevocationRequest(ServerRequestInterface $request): ResponseInterface
    {
        $postParams = $request->getParsedBody();

        $token = $postParams['token'] ?? null;
        $tokenHint = $postParams['token_type_hint'] ?? null;

        if (null === $token || null === $tokenHint) {
            throw OAuth2Exception::invalidRequest(
                'Cannot revoke a token as the "token" and/or "token_type_hint" parameters are missing'
            );
        }

        if ($tokenHint !== 'access_token' && $tokenHint !== 'refresh_token') {
            throw OAuth2Exception::unsupportedTokenType(sprintf(
                'Authorization server does not support revocation of token of type "%s"',
                $tokenHint
            ));
        }

        if ($tokenHint === 'access_token') {
            $token = $this->accessTokenService->getToken((string) $token);
        } else {
            $token = $this->refreshTokenService->getToken((string) $token);
        }

        $response = new Response();

        // According to spec, we should return 200 if token is invalid
        if (null === $token) {
            return $response;
        }

        // Now, we must validate the client if the token was generated against a non-public client
        if (null !== $token->getClient() && ! $token->getClient()->isPublic()) {
            $requestClient = $this->getClient($request, false);

            if ($requestClient !== $token->getClient()) {
                throw OAuth2Exception::invalidClient('Token was issued for another client and cannot be revoked');
            }
        }

        try {
            if ($tokenHint === 'access_token') {
                $this->accessTokenService->deleteToken($token);
            } else {
                $this->refreshTokenService->deleteToken($token);
            }
        } catch (Throwable $exception) {
            // According to spec (https://tools.ietf.org/html/rfc7009#section-2.2.1), we should return a server 503
            // error if we cannot delete the token for any reason
            $response = $response->withStatus(503, 'An error occurred while trying to delete the token');
        }

        return $response;
    }

    /**
     * Get the client (after authenticating it)
     *
     * According to the spec (http://tools.ietf.org/html/rfc6749#section-2.3), for public clients we do
     * not need to authenticate them
     *
     * @throws OAuth2Exception (invalid_client) When a client secret is missing or client authentication failed
     */
    private function getClient(ServerRequestInterface $request, bool $allowPublicClients): ?Client
    {
        list($id, $secret) = $this->extractClientCredentials($request);

        // If the grant type we are issuing does not allow public clients, and that the secret is
        // missing, then we have an error...
        if (! $allowPublicClients && ! $secret) {
            throw OAuth2Exception::invalidClient('Client secret is missing');
        }

        // If we allow public clients and no client id was set, we can return null
        if ($allowPublicClients && ! $id) {
            return null;
        }

        $client = $this->clientService->getClient($id);

        // We delegate all the checks to the client service
        if (null === $client || (! $allowPublicClients && ! $client->authenticate($secret))) {
            throw OAuth2Exception::invalidClient('Client authentication failed');
        }

        return $client;
    }

    /**
     * Create a response from the exception, using the format of the spec
     *
     * @link   http://tools.ietf.org/html/rfc6749#section-5.2
     */
    private function createResponseFromOAuthException(OAuth2Exception $exception): ResponseInterface
    {
        $payload = [
            'error' => $exception->getCode(),
            'error_description' => $exception->getMessage(),
        ];

        return new Response\JsonResponse($payload, 400);
    }

    /**
     * Extract the client credentials from Authorization header or POST data
     */
    private function extractClientCredentials(ServerRequestInterface $request): array
    {
        // We first try to get the Authorization header, as this is the recommended way according to the spec
        if ($request->hasHeader('Authorization')) {
            // The value is "Basic xxx", we are interested in the last part
            $parts = explode(' ', $request->getHeaderLine('Authorization'));
            $value = base64_decode(end($parts));

            list($id, $secret) = explode(':', $value);
        } else {
            $postParams = $request->getParsedBody();

            $id = $postParams['client_id'] ?? null;
            $secret = $postParams['client_secret'] ?? null;
        }

        return [$id, $secret];
    }
}
